cmd/snap-seccomp-blacklist/snap-seccomp-blacklist.c

```diff
@@ -0,0 +1,215 @@
+    sc_err = seccomp_rule_add(ctx, SCMP_ACT_ERRNO(EPERM), sys_ioctl_nr, 1, no_tty_inject);
+
+    if (sc_err < 0) {
+        showerr("cannot add rule preventing the use high bits in ioctl");
+        return sc_err;
+    }
+    return 0;
+}
+
+typedef struct arch_set {
+    const char *name;
+    const uint32_t *arch_tags;
+    size_t num_arch_tags;
+} arch_set;
+
+int main(int argc, char **argv) {
+    const uint32_t le_arch_tags[] = {
+        SCMP_ARCH_X86_64,
+        SCMP_ARCH_AARCH64,
+        SCMP_ARCH_PPC64LE,
+        SCMP_ARCH_S390X,
+    };
+    const uint32_t be_arch_tags[] = {
+        SCMP_ARCH_S390X,
+    };
+    const arch_set arch_sets[] = {
+        {"LE", le_arch_tags, sizeof le_arch_tags / sizeof *le_arch_tags},
+        {"BE", be_arch_tags, sizeof be_arch_tags / sizeof *be_arch_tags},
+    };
+    int rc = -1;
+
+    for (size_t i = 0; i < sizeof arch_sets / sizeof *arch_sets; ++i) {
+        const arch_set *as = &arch_sets[i];
+        int sc_err;
+        int fd = -1;
+        int fname_len;
+        char fname[PATH_MAX];
+
+        scmp_filter_ctx ctx = NULL;
+        ctx = seccomp_init(SCMP_ACT_ALLOW);
+        if (ctx == NULL) {
+            showerr("cannot construct seccomp context");
+            return -rc;
+        }
+        sc_err = populate_filter(ctx, as->arch_tags, as->num_arch_tags);
+        if (sc_err < 0) {
+            seccomp_release(ctx);
+            return -rc;
+        }
+
+        /* Save pseudo-code program */
+        fname_len = snprintf(fname, sizeof fname, "%s-blacklist.pfc", as->name);
+        if (fname_len < 0 || fname_len >= sizeof fname) {
+            showerr("cannot format file name");
+            seccomp_release(ctx);
+            return -rc;
+        }
+        fd = open(fname, O_CREAT | O_TRUNC | O_WRONLY | O_NOFOLLOW, 0644);
+        if (fd < 0) {
+            showerr("cannot open %s", fname);
+            seccomp_release(ctx);
+            return -rc;
+        }
+        sc_err = seccomp_export_pfc(ctx, fd);
+        if (sc_err < 0) {
+            showerr("cannot export PFC program");
+            seccomp_release(ctx);
+            close(fd);
+            return -rc;
+        }
+
+        close(fd);
+
+        /* Save binary program. */
+        fname_len = snprintf(fname, sizeof fname, "%s-blacklist.bpf", as->name);
+        if (fname_len < 0 || fname_len >= sizeof fname) {
+            showerr("cannot format file name");
+            seccomp_release(ctx);
+            return -rc;
+        }
+        fd = open(fname, O_CREAT | O_TRUNC | O_WRONLY | O_NOFOLLOW, 0644);
+        if (fd < 0) {
+            showerr("cannot open %s", fname);
+            seccomp_release(ctx);
+            return -rc;
+        }
+        sc_err = seccomp_export_bpf(ctx, fd);
+        if (sc_err < 0) {
+            showerr("cannot export BPF program");
+            seccomp_release(ctx);
+            close(fd);
+            return -rc;
+        }
+
+        close(fd);
+        seccomp_release(ctx);
+    }
```

Comment:
Question: we may succesfully generate bpf for the first set and for some reason fail on the second set; is this ok to have just one bpf file or should it be all-or-nothing?

Reply:
Ah, I suppose it's fine since those are inputs during the build, and we will simply fail in the Makefile.

Reply:
In addition make is smart enough to unlink partially created target files.